test/Microsoft.AspNetCore.SignalR.Common.Tests/Internal/Protocol/Utf8BufferTextWriterTests.cs

```diff
@@ -0,0 +1,135 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.Text;
+using Microsoft.AspNetCore.SignalR.Internal.Protocol;
+using Xunit;
+
+namespace Microsoft.AspNetCore.SignalR.Common.Tests.Internal.Protocol
+{
+    public class Utf8BufferTextWriterTests
+    {
+        public class TestBufferWriter : IBufferWriter<byte>
+        {
+            public Memory<byte> Buffer { get; set; }
+            public int Position { get; set; }
+
+            public TestBufferWriter(int bufferSize)
+            {
+                Buffer = new Memory<byte>(new byte[bufferSize]);
+            }
+
+            public void Advance(int count)
+            {
+                Position += count;
+            }
+
+            public Memory<byte> GetMemory(int sizeHint = 0)
+            {
+                if (sizeHint == 0)
+                {
+                    // return remaining
+                    sizeHint = Buffer.Length - Position;
+                }
+
+                return Buffer.Slice(Position, sizeHint);
+            }
+
+            public Span<byte> GetSpan(int sizeHint = 0)
+            {
+                return GetMemory(sizeHint).Span;
+            }
+        }
+
+        [Fact]
+        public void WriteChar_Unicode()
+        {
+            TestBufferWriter bufferWriter = new TestBufferWriter(4096);
+            Utf8BufferTextWriter textWriter = new Utf8BufferTextWriter(bufferWriter);
+
+            textWriter.Write('[');
+            Assert.Equal(1, bufferWriter.Position);
+            Assert.Equal((byte)'[', bufferWriter.Buffer.Span[0]);
+
+            textWriter.Write('"');
+            Assert.Equal(2, bufferWriter.Position);
+            Assert.Equal((byte)'"', bufferWriter.Buffer.Span[1]);
+
+            textWriter.Write('\u00A3');
+            Assert.Equal(2, bufferWriter.Position);
```

Comment:
Why is it still 2 after writing here?

Reply:
Because the char is in the char buffer. Multiple calls will fill up that buffer so they are written at once instead of encoding char by char.

But it just occurred to me that I should ensure that they get written on a flush or close on the reader.